x/ccv/consumer/keeper/soft_opt_out.go

```diff
@@ -0,0 +1,59 @@
+package keeper
+
+import (
+	"encoding/binary"
+	"sort"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/interchain-security/x/ccv/consumer/types"
+)
+
+// SetSmallestNonOptOutPower sets the smallest validator power that cannot soft opt out.
+func (k Keeper) SetSmallestNonOptOutPower(ctx sdk.Context, power uint64) {
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.SmallestNonOptOutPowerKey(), sdk.Uint64ToBigEndian(power))
+}
+
+// UpdateSmallestNonOptOutPower updates the smallest validator power that cannot soft opt out.
+// This is the smallest validator power such that the sum of the power of all validators with a lower power
+// is less than [SoftOptOutThreshold] of the total power of all validators.
+func (k Keeper) UpdateSmallestNonOptOutPower(ctx sdk.Context) {
+	// get all validators
+	valset := k.GetAllCCValidator(ctx)
+
+	// sort validators by power ascending
```

Comment:
if valset empty we return. This only happens in tests, add warning log

Reply:
https://github.com/cosmos/interchain-security/pull/833/commits/741da8a570eb6c0e7ee34f61280cd5dd4ce37758